Java/core/src/main/java/com/amazon/randomcutforest/tree/AbstractNodeStore.java

```diff
@@ -0,0 +1,527 @@
+    }
+
+    public double[] scoreScalar(int node, int depth, BoundingBoxFloat box, int ignoreMass, float[] point,
+            IPointStoreView<float[]> pointStoreView, BiFunction<Double, Double, Double> scoreSeen,
+            BiFunction<Double, Double, Double> scoreUnseen, Function<Double, Double> treeDamp) {
+        if (isLeaf(node)) {
+            double mass = getMass(node);
+            float[] oldPoint = pointStoreView.get(getPointIndex(node));
+            int ignoreFlag = (mass > ignoreMass) ? 1 : 0;
+            if (box != null) {
+                box.replaceBox(oldPoint);
+            }
+            if (Arrays.equals(point, oldPoint) && ignoreFlag == 1) {
+                return new double[] { 0.0, treeDamp.apply(mass) * scoreSeen.apply(depth * 1.0, mass), ignoreFlag };
+            } else {
+                return new double[] { 1.0, scoreUnseen.apply(1.0 * depth, mass), ignoreFlag };
+            }
+        }
+        double[] answer;
+        int sibling;
+        if (leftOf(node, point)) {
+            answer = scoreScalar(getLeftIndex(node), depth + 1, box, ignoreMass, point, pointStoreView, scoreSeen,
+                    scoreUnseen, treeDamp);
+            if (answer[0] != 0.0 && box != null) {
+                if (answer[2] == 1) {
+                    growNodeBox(box, pointStoreView, node, getRightIndex(node));
+                } else {
+                    box.copyFrom(getBox(getRightIndex(node), pointStoreView));
+                    answer[2] = 1;
+                }
+            }
+        } else {
+            answer = scoreScalar(getRightIndex(node), depth + 1, box, ignoreMass, point, pointStoreView, scoreSeen,
+                    scoreUnseen, treeDamp);
+            if (answer[0] != 0.0 && box != null) {
+                if (answer[2] == 1) {
+                    growNodeBox(box, pointStoreView, node, getLeftIndex(node));
+                } else {
+                    box.copyFrom(getBox(getLeftIndex(node), pointStoreView));
+                    answer[2] = 1;
+                }
+            }
+        }
+
+        if (answer[0] == 0.0) {
+            return answer;
+        }
+
+        double prob = (ignoreMass == 0) ? probabilityOfCut(node, point, pointStoreView, box)
+                : box.probabilityOfCut(point);
+        answer[0] = prob;
+        answer[1] = answer[1] * (1.0 - prob) + prob * scoreUnseen.apply(1.0 * depth, 1.0 * getMass(node));
+        return answer;
+    }
+
+    public abstract int getCutDimension(int index);
+
+    public double getCutValue(int index) {
+        return cutValue[index - 1];
+    }
+
+    public double getNodeCacheFraction() {
+        return nodeCacheFraction;
+    }
+
+    protected <R> void traversePathToLeafAndVisitNodes(double[] point, Visitor<R> visitor, int root,
+            IPointStoreView<float[]> pointStoreView, Function<double[], double[]> projectToTree) {
+        AbstractNodeView currentNodeView = new AbstractNodeView(this, pointStoreView, root);
```

Comment:
Is AbstractNodeView an abstract class? If it is, I feel like we shouldn't be able to construct one like this.

Reply:
No ... bad naming. Will change.